api-select.go

```diff
@@ -0,0 +1,621 @@
+	lenOfTrim := 0
+	for {
+		// select block for catching a user request for closing.
+		select {
+		// Logic is basically that the .Close() method will close the streamer.lock
+		// channel, Therefore if Ok is false(the channel is closed) then we can
+		// close our response channel and return. The .Close() method blocks on the
+		// streamer.Response channel being closed, so that it is guaranteed that this
+		// method has exited since it is a defered close.
+		case _, ok := <-streamer.lock:
+			if !ok {
+				defer close(streamer.Response)
+				return
+			}
+		default:
+			var prelude preludeInfo
+			var headers *headerInfo
+			// Create CRC code
+			crc := crc32.New(crc32.IEEETable)
+			crcReader := io.TeeReader(message, crc)
+			var err error
+			var payload []byte
+			// Extract the prelude(12 bytes) into a struct to extract relevant
+			// information.
+			prelude, err = processPrelude(crcReader, crc)
+			if err != nil {
+				if streamer.err == nil {
+					streamer.err = err
+				}
+				close(streamer.Response)
+				return
+			}
+			// Extract the headers(variable bytes) into a struct to extract relevant
+			// information
+			if prelude.headerLen > 0 {
+				headers, err = extractHeader(io.LimitReader(crcReader, int64(prelude.headerLen)))
+				if err != nil {
+					if streamer.err == nil {
+						streamer.err = err
+					}
+					close(streamer.Response)
+					return
+				}
+			}
+			// Calculate the length of the payload so that the appropriate amount of
+			// bytes can be read
+			payloadLen := calcPayloadLen(prelude)
+			// Checks that there is a payload and if there is, then extracts it.
+			if payloadLen > 0 {
+				payload, err = extractPayload(io.LimitReader(crcReader, int64(payloadLen)), payloadLen)
+				if err != nil {
+					if streamer.err == nil {
+						streamer.err = err
+					}
+					close(streamer.Response)
+					return
+				}
+				// Special case for dealing with messages that are partial.
+				var state bool
+				for isPartial(payload) {
+					// state indicates whether it is a partial record or not, so if it
+					// is false then only processPartial should be called
+					payload, partialRecordVal, lenOfTrim = processPartial(payload)
+					// sends complete records to client, while partial records are
+					// stitched.
+					processEvent(payload, headers, streamer)
+					// return the complete part of the message so far to the client
+					// for loop in case there are multiple sequential partial rows
+					// Extracts the partial payload.
+					myPayload, err := extractPayload(io.LimitReader(crcReader, int64(lenOfTrim)), lenOfTrim)
+					if err != nil {
+						if streamer.err == nil {
+							streamer.err = err
+						}
+						close(streamer.Response)
+						return
+					}
+					// knits together the payload with the previous partial one.
+					payload = append(partialRecordVal, myPayload...)
+					// update state to keep track of whether partial records were
+					// processed
+					state = !isPartial(payload)
```

Comment:
This state variable is not needed. As written, the for loop is exited when the isPartial condition evaluates to false. So processEvent must be called every time after the for loop.

Reply:
The state variable is actually needed, because in the case that there is NOT a partial record sent, e.g it just one complete record, you want to make sure that the processEvent immediately following the isPartial, is not called. Also lets say there are 2 parts to a partial record, after the 2nd one, the record is now complete, so then we update the state and quit the for loop and send the processEvent. The reason this processEvent has to be here is because when partial records are sent only the first one follows the format of a normal message e.g prelude header payload crc, the rest follow just a format of .. payload .., so thats why it is written like this